### PR TITLE
Handle non-Countables passed to count

### LINKED ISSUE #7229: Incorrect behavior of CountOnNullRector
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-main      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See [https://getrector.org/demo/3be70b11-6e9a-45a4-b3f7-d46494ccf07e](https://getrector.org/demo/3be70b11-6e9a-45a4-b3f7-d46494ccf07e)

```php
<?php

final class DemoFile
{
    public function run()
    {
        $my_var= "abc";
        if (strlen("There is a condition that is true only sometimes") == rand(5, 50)) {
            $my_var= [1,2];
        }
        $count = count($my_var);
        if ($count) {
            echo $count . " items found";
        } else {
            echo "No items found";
        }
    }
}
```

### Responsible rules

* `CountOnNullRector`

## Observed Behaviour
```
-        $count = count($my_var);
+        $count = is_countable($my_var) ? count($my_var) : 0;
```

## What's going wrong
[Prior to PHP 8, if `count` was called with a parameter that wasn't an array, or an object that implemented the Countable interface, it would return 1.](https://www.php.net/manual/en/function.count.php#:~:text=Prior%20to%20PHP%208.0.0%2C%20if%20the%20parameter%20was%20neither%20an%20array%20nor%20an%20object%20that%20implements%20the%20Countable%20interface%2C%201%20would%20be%20returned%2C%20unless%20value%20was%20null%2C%20in%20which%20case%200%20would%20be%20returned)
This results in a change in the value of `$count` when `$my_var` it is a string:
```
Prior to PHP 8 (with no rector changes):
$count = 1

After PHP 8 (with no rector changes):
$count = Error

With rector changes:
$count = 0
```

## Expected Behavior
`$count` should be `1`, when `$my_var` it is a string

**Note**: I know that this is a PHP 7.1 rector, but since there is no new rector to fix this, I'm using this rector to report the bug.

Since the change by rector only takes into account arrays/Countables/null, if a non-Countable is passed, 0 will be returned now. This changes the behaviour/logic of legacy code. 

The solution/correct change would be:
```
-        $count = count($my_var);
+        $count = is_countable($my_var) ? count($my_var) : (is_null($my_var) ? 0 : 1);
```

### PR DESCRIPTION
**Pull request by brondibur**
Fixes [#7229](https://github.com/rectorphp/rector/issues/7229)

### CODE REVIEW

**Comment by samsonasik** [comment]
it seems make test failure: https://github.com/rectorphp/rector-src/runs/6845274833?check_suite_focus=true

 also please add new test case with expected output at https://github.com/rectorphp/rector-src/tree/main/rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture

**Comment by brondibur** [comment]
@samsonasik Fixed test failures. Have updated the existing test cases for the rector I've updated.

**Comment by samsonasik** [comment]
I think this behavior should not be applied by default, probably make configurable, eg: add configurable `handleCountOnString` which default to false, so can be enabled when needed.

**Comment by brondibur** [comment]
@samsonasik Why do you think it shouldn't be the default behaviour? 
The existing `CountOnNullRector` is obsolete, and the fixes provided by this rector are incorrect from the time PHP 8.0.0 was launched since it clearly changes behaviour of legacy code as shown here: https://github.com/rectorphp/rector/issues/7229

Since the idea behind rector is to make code compatible with newer versions, while maintaining existing behaviour, this should certainly be the default behaviour.

**Comment by samsonasik** [comment]
Because I think the change in the fixtures are not relevant, another solution is to add Union check which the type can possibly string

**Comment by brondibur** [comment]
@samsonasik Why do you think it's not relevant?

I guess you didn't understand what this change is about.
`CountOnNullRector` is currently changing the behaviour of legacy code in case someone uses Rector to upgrade to PHP versions 8.0.0 or higher.
And in case the upgrade is to a lower PHP version, the change is not future proof.

Also, this is not limited to `string`s. This involves every parameter passed that doesn't implement the `Countable` interface. Please go through the [linked issue](https://github.com/rectorphp/rector/issues/7229).

**Comment by samsonasik** [comment]
It needs to be skipped when value type is obvious, only allow transform to add `is_null($values) ? 0 : 1` when it is `UnionType` which has possible `NullType`.

Too much transformation when value type is obvious will make the rule unusable.

**Comment by brondibur** [comment]
@samsonasik Looks like you still haven't understood what this PR does. This change isn't checking for `null` types, the original `CountOnNullRector` already did that pretty well. But that Rector's fixes became obsolete since the launch of PHP 8.0.0. 
This change is checking for non-`Countable`s, thereby expanding `CountOnNullRector` and making sure it delivers the correct, and future-proof fixes while maintaining legacy behaviour since the value type can be `Nullable` even if it isn't a `Countable`. 

**How, with an example:**
Original code:
```
$sum = count($var);
```
New code:
```
$sum = is_countable($var) ? count($var) : (is_null($var) ? 0 : 1);
```


### Breaking it down:
**If >> $var is an array or implements the Countable interface:**
```
is_countable($var) 
```
**Then >> Return $var's count:**
```
count($var)
```
**Else >> $var is not an array and doesn't implement the Countable interface:**
```
(is_null($var) ? 0 : 1)
```
&nbsp;&nbsp;&nbsp;&nbsp;**If >> $var is null:**
```
is_null($var)
```
&nbsp;&nbsp;&nbsp;&nbsp;**Then Return zero (0):**
```
0
```
&nbsp;&nbsp;&nbsp;&nbsp;**Else >> $var is not an array, doesn't implement the Countable interface, and is also not null. So return one (1):**
```
1
```

File: rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/false_true_class.php.inc [hunk 1]
```diff
@@ -27,13 +27,13 @@ use stdClass;
 function countables2()
 {
     $values = false;
-    $count = is_array($values) || $values instanceof \Countable ? count($values) : 0;
+    $count = is_array($values) || $values instanceof \Countable ? count($values) : (is_null($values) ? 0 : 1);
```
Comment (samsonasik) [inline rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/false_true_class.php.inc:30]: false is a constant value, I think replace it directly to 1 should be the way to go, so should be:

```suggestion
    $count = 1;
```

other fixture should applies to for constant value as well.

File: rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/local_property_with_doc.php.inc [hunk 2]
```diff
@@ -32,7 +32,7 @@ final class LocalPropertyWithDoc
     public function getTitle()
     {
         $this->titles = null;
-        $titleCount = $this->titles === null ? 0 : count($this->titles);
+        $titleCount = is_array($this->titles) || $this->titles instanceof \Countable ? count($this->titles) : (is_null($this->titles) ? 0 : 1);
```
Comment (samsonasik) [inline rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/local_property_with_doc.php.inc:35]: it should be directly assigned to 0 as passed constant null value

```suggestion
        $titleCount = 0;
```

**Comment by samsonasik** [comment]
On `MixedType`, transformation is ok, but for `UnionType` with obvious types or obvious constant value type, I think it should be changed per my comments on the fixtures.

By that, we can have eliminate unnecessary transformation.

File: rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/type_changed_on_if.php.inc [hunk 3]
```diff
@@ -30,7 +30,7 @@ final class TypeChangedOnIf
             $data = new \stdClass;
         }
 
-        $data[] = (is_array($data) || $data instanceof \Countable ? count($data) : 0) !== 0 ? 'a' : 'b';
+        $data[] = (is_array($data) || $data instanceof \Countable ? count($data) : (is_null($data) ? 0 : 1)) !== 0 ? 'a' : 'b';
```
Comment (samsonasik) [inline rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/type_changed_on_if.php.inc:33]:
```suggestion
        $data[] = (is_array($data) ? count($data) : 1) !== 0 ? 'a' : 'b';

**Comment by TomasVotruba** [comment]
I'm wondering, what is missing here to finish?

**Comment by samsonasik** [comment]
This PR change cause too many changes for obvious constant value type, which make the rule is unusable. 

Also, I think the fallback to 0 is mostly a trade off for verify value in count() usage, which should be as default value in case it got invalid result.

**Comment by TomasVotruba** [comment]
What would you suggest? Keep rule as it is?

**Comment by TomasVotruba** [comment]
Ref https://github.com/rectorphp/rector/issues/7229

**Comment by samsonasik** [comment]
Yes, keep it as it current behaviour.

**Comment by TomasVotruba** [comment]
Closing then, see https://github.com/rectorphp/rector/issues/7229

**Pull request closed without merging**